soc/arm/nxp_imx/rt/power_rt11xx.c

```diff
@@ -0,0 +1,337 @@
+	dcdc_setpoint_config.buckVDD1P0TargetVoltage = buck1_0_voltage;
+	dcdc_setpoint_config.standbyVDD1P8TargetVoltage = standby1_8_voltage;
+	dcdc_setpoint_config.standbyVDD1P0TargetVoltage = standby1_0_voltage;
+	DCDC_SetPointInit(DCDC, &dcdc_setpoint_config);
+
+	DCDC_GetDefaultConfig(&dcdc_config);
+	dcdc_config.controlMode = kDCDC_SetPointControl;
+	DCDC_Init(DCDC, &dcdc_config);
+}
+
+static void system_enter_sleep(gpc_cpu_mode_t gpc_mode)
+{
+	__ASSERT_NO_MSG(gpc_mode != kGPC_RunMode);
+
+	if (gpc_mode == kGPC_WaitMode) {
+		/* Clear SLEEPDEEP bit to enter WAIT mode*/
+		SCB->SCR &= ~SCB_SCR_SLEEPDEEP_Msk;
+	} else {
+		/* Set SLEEPDEEP bit to enter STOP mode */
+		SCB->SCR |= SCB_SCR_SLEEPDEEP_Msk;
+	}
+	/* FIXME: When this function is entered the Kernel has disabled
+	 * interrupts using BASEPRI register. This is incorrect as it prevents
+	 * waking up from any interrupt which priority is not 0. Work around the
+	 * issue and disable interrupts using PRIMASK register as recommended
+	 * by ARM.
+	 */
+
+	/* Set PRIMASK */
+	__disable_irq();
+	/* Set BASEPRI to 0 */
+	irq_unlock(0);
+
+	/* WFI instruction will start entry into WAIT/STOP mode */
+	LOG_DBG("Entering LPM via WFI");
+	__WFI();
+}
+
+void cpu_mode_transition(gpc_cpu_mode_t mode, bool enable_standby)
+{
+	GPC_CM_SetNextCpuMode(GPC_CPU_MODE_CTRL, mode);
+	GPC_CM_EnableCpuSleepHold(GPC_CPU_MODE_CTRL, true);
+
+	/* Mask debugger wakeup */
+	GPC_CPU_MODE_CTRL->CM_NON_IRQ_WAKEUP_MASK |=
+		GPC_CPU_MODE_CTRL_CM_NON_IRQ_WAKEUP_MASK_EVENT_WAKEUP_MASK_MASK |
+		GPC_CPU_MODE_CTRL_CM_NON_IRQ_WAKEUP_MASK_DEBUG_WAKEUP_MASK_MASK;
+
+	if (enable_standby) {
+		/* Set standby request */
+		GPC_CM_RequestStandbyMode(GPC_CPU_MODE_CTRL, mode);
+	} else {
+		/* Clear standby request */
+		GPC_CM_ClearStandbyModeRequest(GPC_CPU_MODE_CTRL, mode);
+	}
+
+	/* Execute WFI- GPC will receive sleep request from CPU */
+	system_enter_sleep(mode);
+}
+
+/**
+ * SOC specific low power mode implementation
+ * Drop to lowest power state possible given system's request
+ */
+__weak void pm_power_state_set(struct pm_state_info info)
+{
+	/* Extract set point and GPC mode from the substate ID */
+	uint8_t set_point = IMX_SPC(info.substate_id);
+	gpc_cpu_mode_t gpc_mode = IMX_GPC_MODE(info.substate_id);
+	uint8_t current_set_point = GPC_SP_GetCurrentSetPoint(GPC_SET_POINT_CTRL);
+
+	LOG_DBG("Switch to Set Point %d, GPC Mode %d requested", set_point, gpc_mode);
+	if (gpc_mode != kGPC_RunMode && (current_set_point != set_point)) {
+		/* Request set point transition at sleep */
+		GPC_CM_RequestSleepModeSetPointTransition(GPC_CPU_MODE_CTRL,
+			set_point, set_point, kGPC_CM_RequestPreviousSetpoint);
+		cpu_mode_transition(gpc_mode, true);
+	} else if (gpc_mode != kGPC_RunMode) {
+		/* Request CPU mode transition without set mode transition */
+		GPC_CM_RequestRunModeSetPointTransition(GPC_CPU_MODE_CTRL,
+							current_set_point);
+		cpu_mode_transition(gpc_mode, true);
+	} else if (current_set_point != set_point) {
```

Comment:
This should not be necessary, the subsystem only calls this function from the
active state. In other words, will never be called from a sleep state.

Reply:
Thanks for the heads up- I'll remove this block